src/replication.c

```diff
@@ -197,27 +202,207 @@ void feedReplicationBacklogWithObject(robj *o) {
+            /* Use last not full block. */
+            if (used_last_block) {
+                slave->used_blocks_of_repl_buf++;
+                slave->used_size_of_repl_buf += used_last_block->size;
+            }
+        }
+
+        /* Incr used replication buffer size, check output buffer limit
+         * only when add new block. */
+        if (add_new_block) {
+            slave->used_blocks_of_repl_buf++;
+            slave->used_size_of_repl_buf += tail->size;
+            closeClientOnOutputBufferLimitReached(slave, 1);
+        }
+    }
+
+    /* For replication backlog */
+    if (server.repl_backlog->ref_repl_buf_node == NULL) {
+        server.repl_backlog->ref_repl_buf_node = start_node;
+        /* Only increase the start block reference count. */
+        ((replBufBlock *)listNodeValue(start_node))->refcount++;
+
+        /* Replication buffer must be empty before adding replication stream
+         * into replication backlog. */
+        serverAssert(used_last_block == NULL);
+    }
+    if (add_new_block) {
+        /* To make search offset from replication buffer blocks quickly
+         * when replicas ask partial resynchronization, we record one block
+         * every 64 blocks. */
+        server.repl_backlog->unrecorded_count++;
+        if (server.repl_backlog->unrecorded_count >= RECORD_ONE_EVERY_BLOCKS) {
+            uint64_t encoded_offset = htonu64(tail->repl_offset);
+            raxInsert(server.repl_backlog->blocks_index,
+                (unsigned char*)&encoded_offset, sizeof(uint64_t),
+                listLast(server.repl_buffer_blocks), NULL);
+            server.repl_backlog->unrecorded_count = 0;
+        }
+    }
+    /* Trim replication backlog. */
+    incrementalTrimReplicationBacklog(TRIM_REPL_BUF_BLOCKS_PER);
+}
+
+/* Free replication buffer blocks that are referenced by this client. */
+void freeSlaveReferencedReplBuffer(client *replica) {
+    if (replica->ref_repl_buf_node != NULL) {
+        /* Decrease the start buffer node reference count. */
+        replBufBlock *o = listNodeValue(replica->ref_repl_buf_node);
+        serverAssert(o->refcount > 0);
+        o->refcount--;
+        incrementalTrimReplicationBacklog(TRIM_REPL_BUF_BLOCKS_PER);
+    }
+    replica->ref_repl_buf_node = NULL;
+    replica->ref_block_pos = 0;
+    replica->used_blocks_of_repl_buf = 0;
+    replica->used_size_of_repl_buf = 0;
 }
 
-/* Propagate write commands to slaves, and populate the replication backlog
- * as well. This function is used if the instance is a master: we use
- * the commands received by our clients in order to create the replication
- * stream. Instead if the instance is a slave and has sub-slaves attached,
- * we use replicationFeedSlavesFromMasterStream() */
+/* Propagate write commands to replication buffer. Now, we uniform slaves's
+ * output buffer and replication backlog into one shared replication buffer.
+ * Replication backlog just is logical.
```

Comment:
Wanted to remove the word "slave" but actually, I don't think the rest of the text is needed here.
instead of talking about slaves and backlog, we can say this function manages the replication stream (a term that was already used here).
```suggestion
/* Propagate write commands to replication stream.
```
Actually, this line can be completely removed, i think the rest of the comment stands for itself.